db/version_set.cc

```diff
@@ -1124,67 +1244,172 @@ void LevelIterator::Seek(const Slice& target) {
+      // In SeekForPrev case, it is possible that the target is less than file's
+      // lower boundary since largest key is used to determine file index. When
+      // target is less than file's lower boundary, sentinel key should not be
+      // set so that SeekForPrev does not result in a key larger than target.
+      // This is correct in that there is no need to keep the range tombstones
+      // in this file alive as they only cover keys starting from lower
+      // boundary, which is after `target`.
+      TrySetDeleteRangeSentinel(file_smallest_key(file_index_),
+                                true /* try_prefix_sentinel */);
+    }
+    if (!to_return_sentinel_) {
+      SkipEmptyFileBackward();
+    }
   }
   CheckMayBeOutOfLowerBound();
 }
 
 void LevelIterator::SeekToFirst() {
+  ClearSentinel();
   InitFileIterator(0);
   if (file_iter_.iter() != nullptr) {
     file_iter_.SeekToFirst();
+    if (range_tombstone_iter_) {
+      // We do this in SeekToFirst() and SeekToLast() since
+      // we could have an empty file with only range tombstones.
+      TrySetDeleteRangeSentinel(file_largest_key(file_index_));
+    }
+  }
+  if (!to_return_sentinel_) {
+    SkipEmptyFileForward();
   }
-  SkipEmptyFileForward();
   CheckMayBeOutOfLowerBound();
 }
 
 void LevelIterator::SeekToLast() {
+  ClearSentinel();
   InitFileIterator(flevel_->num_files - 1);
   if (file_iter_.iter() != nullptr) {
     file_iter_.SeekToLast();
+    if (range_tombstone_iter_) {
+      TrySetDeleteRangeSentinel(file_smallest_key(file_index_));
+    }
+  }
+  if (!to_return_sentinel_) {
+    SkipEmptyFileBackward();
   }
-  SkipEmptyFileBackward();
   CheckMayBeOutOfLowerBound();
 }
 
 void LevelIterator::Next() {
+  // For range tombstones, Next() call should update range_tombstone_iter_
+  // if file_iter_ changes. Maintain active_range_tombstones_ set
+  // correspondingly. Before advancing to next SST file, check boundary for
+  // possible tombstone sentinel key.
   assert(Valid());
-  file_iter_.Next();
-  SkipEmptyFileForward();
+  if (to_return_sentinel_) {
+    // file_iter_ was at end of file already if we were to return sentinel.
+    ClearSentinel();
+    SkipEmptyFileForward();
+  } else {
+    file_iter_.Next();
+    if (range_tombstone_iter_) {
+      TrySetDeleteRangeSentinel(file_largest_key(file_index_));
+    }
+    if (!to_return_sentinel_) {
+      SkipEmptyFileForward();
+    }
+  }
 }
 
 bool LevelIterator::NextAndGetResult(IterateResult* result) {
   assert(Valid());
-  bool is_valid = file_iter_.NextAndGetResult(result);
+  // file_iter_ is at EOF already when to_return_sentinel_
+  bool is_valid = !to_return_sentinel_ && file_iter_.NextAndGetResult(result);
+  bool prefix_sentinel = false;
   if (!is_valid) {
+    if (to_return_sentinel_) {
+      prefix_sentinel = prefix_sentinel_;
+      ClearSentinel();
+    } else if (range_tombstone_iter_) {
+      TrySetDeleteRangeSentinel(file_largest_key(file_index_));
```

Comment:
Would it simplify things if SkipEmptyFileForward() paused at the sentinel? Seems there's some redundancy injecting sentinel after passing last point key in the file (here), and injecting sentinel for a file containing no point keys (SkipEmptyFileForward()). Although the function name might need to change.

Reply:
Yes, it simplifies many call sites of SkipEmptyFile*(). Updated the PR with this change.